Correctly rounded floating point `div_euclid`.

tczajka opened the pull request:
Fixes #107904.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

tczajka committed:
Correctly rounded floating point `div_euclid`.

Fixes #107904.

tgross35 commented:
At a surface level, I'm wondering if these would be better in libm so we get the more full test suite against mpfr.

tczajka commented:
> At a surface level, I'm wondering if these would be better in libm so we get the more full test suite against mpfr.

Makes sense, but would it be reasonable for `std` to depend on `libm` to get this one function?

tgross35 commented:
We already ship `libm` with the sysroot, so this would just be via an extern symbol like the other functions that call `sys::cmath`. I think this is fine for now though, that can just be something to investigate in the future (considering we have some ongoing changes w.r.t. when `libm` symbols get made available).

tczajka commented:
> We already ship `libm` with the sysroot as part of `compiler_builtins`, so this would just be via an extern symbol like the other functions that call `sys::cmath`. I think this is fine for now though, that can just be something to investigate in the future (considering we have some ongoing changes w.r.t. when `libm` symbols get made available).

Sorry, I deleted a comment you were replying to that asked "is it OK for `std` to depend on `libm` to get this function".

I'm happy to move this to `libm` if it makes sense, although as I understand it `libm` exports C `libm`'s API which doesn't have this function, so that may be a bit confusing.

quaternic commented on library/std/src/f32/div_euclid.rs:
```diff
@@ -0,0 +1,204 @@
+            (true, true) => div_ceil(exp, a_m, b_m),
+        }
+    } else {
+        // `a` or `b` are +-0.0 or infinity or NaN.
+        // `a / b` is also +-0.0 or infinity or NaN.
+        // There is no need to round to an integer.
+        a / b
```
For `finite / inf`, `div_euclid` should give values in `{-1, 0, 1}` depending on the signs, to match both `rem_euclid` and the current behaviour.

tczajka replied:
I'm not convinced `x.div_euclid(inf)` should ever be `-1.0` or `1.0`. I think it's a random artifact of the current naive implementation.

Let's say you have `(-7.0).div_euclid(inf)`. Mathematically it should be `floor(-7 / ∞) = floor(-0) = -0`. I guess one could debate whether `floor(-0.0)` should be `-0.0` or `-1.0`, but `(-0.0f32).floor()` does in fact return `-0.0`, which makes sense given than `(+0.0f32).ceil() == +0.0` .

It can't really be made consistent with `rem_euclid` either way for infinity denominators:
`rem_euclid(-7.0, inf) = inf`
`(-0.0) * inf + inf = NaN + inf = NaN != -7.0`
`(-1.0) * inf + inf = -inf + inf = NaN != -7.0`

tczajka replied:
I do admit there is something pleasant about having `(-7.0).div_euclid(inf) == -1.0`, so I'm open to the possibility. I'm not sure either way. I checked that in Python: `-7.0 // inf` (Python `//` is `div_floor`) gives `-1.0`, so perhaps that makes sense.

But it doesn't quite jive with saying it's `a / b` rounded down (for b > 0), so probably this needs to be clarified in the docs.

`inf` is dumb ;)

tczajka replied:
I definitely don't think it should be NaN as proposed by @traviscross

tczajka replied:
Just because rearranging the formula gives NaN doesn't mean the result must be NaN. The rearranging you did is only necessarily valid for finite non-zero real numbers. By rearranging formulas indiscriminately you can argue anything to be NaN, for example: `1 / 2 = (1 * 0) / (2 * 0) = 0 / 0 = NaN`.

Regarding 0 / inf, I don't see a reason to deviate from what `/` gives:
```
(-0.0).div_euclid(inf) = -0.0
(-0.0).div_euclid(-inf) = +0.0
```

tczajka replied:
Here is a related question: what should `(7.0).rem_euclid(0.0)` be?

It's currently NaN, but according to definitions in Donald Knuth's Concrete Mathematics book, `7 mod 0 = 7` and `-7 mod 0 = -7` (he gives examples like that explicitly).

This also applies to `7.0 % 0.0` and `7u32 % 0u32`.

tczajka replied:
BTW it's bad that `0.0 * inf` is NaN (but that's what IEEE 754 unfortunately mandates). In math it's usually assumed `0 * ∞ = 0` (and that is actually useful for integrals and stuff). With `0 * inf = 0`, `7.0.div_euclid(0.0) = inf` and `7.0.rem_euclid(0.0) = 7` would match:
```
inf * 0 + 7 = 0 + 7 = 7
```

tczajka replied:
It would be useful to have some other source or precedent for what `div_euclid(-7, inf)` means (or equivalently `div_floor(-7, inf)`).

To me there are a few conflicting arguments:
1. `rem_euclid(-7, inf) = inf` does seem to morally indicate that `div_euclid(-7, inf)` should be `-1` not `-0`. Although it's not *totally* clear-cut because that's assuming that `0*inf + (-7) = -7`. Which makes sense, but according to IEEE-754, `0 * inf` is not defined.
2. Python says -1.
3. But the definition "floor(a / b)" indicates -0. So it is hard to explain from first principles why it's -1.

I'm somewhat sympathetic towards the -1 now, partially because of Python. IEEE-754 doesn't provide guidance because it doesn't define this operation.

quaternic replied:
A defining property is that there is a unique integer solution `n` for `x = ny + r, where 0 <= r < |y|`.  For `-|y| <= x < 0`, the solution is `n = -1, r = |y| + x < |y|` regardless of the precise value of `y`. That is, the limit of the quotient is `-1` as `y -> inf`. For infinite divisor the remainder just always "rounds" to infinity in that case.

> 3\. But the definition "floor(a / b)" indicates -0. So it is hard to explain from first principles why it's -1.

Similarly, `lim(y  -> inf) floor(-7 / y) == -1`. You only get 0 as the limit of `-7 / y`.

quaternic replied:
On infinity arithmetic, IEEE-754 says:
```quote
The behavior of infinity in floating-point arithmetic is derived from the limiting cases of real arithmetic
with operands of arbitrarily large magnitude, when such a limit exists. Infinities shall be interpreted in the
affine sense, that is: −∞ < {every finite number} < +∞.
```

tczajka replied:
Yes I buy the limit argument, this makes sense.

tczajka replied:
This is now implemented as -1 or +1 as appropriate.

theemathas commented:
How does this relate to #134062?

tczajka committed:
Fix overflow bug in f16::div_ceil

tczajka commented on library/std/src/f16/div_euclid.rs:
```diff
@@ -0,0 +1,111 @@
+        let aa = u32::from(a) << exp;
+        let bb = u32::from(b);
+        // q <= (2^27 - 2^16) / 2^10 + 1 = 2^17 - 2^6 + 1
+        let q = ((aa - 1) / bb) as u16 + 1;
```
Yes that was a bug in f16::div_ceil, fixed.

tczajka commented on library/std/src/f16/div_euclid.rs:
```diff
@@ -0,0 +1,111 @@
+
+/// Software implementation of `f16::div_euclid`.
+#[allow(dead_code)] // f16::div_euclid is cfg(not(test))
+pub(crate) fn div_euclid(a: f16, b: f16) -> f16 {
+    if let Some((a_neg, a_exp, a_m)) = normal_form(a)
+        && let Some((b_neg, b_exp, b_m)) = normal_form(b)
+    {
+        let exp = a_exp - b_exp;
+        match (a_neg, b_neg) {
+            (false, false) => div_floor(exp, a_m, b_m),
+            (true, false) => -div_ceil(exp, a_m, b_m),
+            (false, true) => -div_floor(exp, a_m, b_m),
+            (true, true) => div_ceil(exp, a_m, b_m),
+        }
+    } else {
+        // `a` or `b` are +-0.0 or infinity or NaN.
+        // `a / b` is also +-0.0 or infinity or NaN.
+        // There is no need to round to an integer.
+        a / b
+    }
+}
```
The correct answer is -inf, this was caused by the overflow bug in `f16::div_ceil`, fixed.

tczajka commented:
I'm now doing some largish refactoring to clean this up a bit, and deal with the `(-finite / inf)` case correctly as suggested by @quaternic. This will mean that this will be done fully via soft-float including special cases.

tczajka committed:
Software `div_euclid` refactoring.

`f{16,32,64,128}/soft.rs` are now very similar. f32 and f64 almost
identical.

`U256` helper moved to `crate::u256`.

`(-5f32).div_euclid(f32::INFINITY)` now returns -1 not 0.

tczajka commented:
OK done refactoring.

tczajka commented:
> It'd need libs-api acceptance, of course, but my feeling is that, whether with or following on to this fix, that we should expose `div_floor`, `div_ceil`, and `div_trunc` as public methods on the floating point types. Perhaps, or perhaps not, that possibility might influence how we think about code organization here.

It's easy to add these using the functions implemented in this PR.

tczajka committed:
div_euclid benchmarks

tczajka commented:
I have added benchmarks for `div_euclid`.

Before the change (incorrectly rounded):
```
    f128::div_euclid_large                                             44859.45/iter +/- 1923.27
    f128::div_euclid_medium                                              135.82/iter    +/- 2.15
    f128::div_euclid_small                                                63.94/iter    +/- 0.27
    f16::div_euclid_large                                                 23.89/iter    +/- 1.24
    f16::div_euclid_small                                                 21.83/iter    +/- 1.35
    f32::div_euclid_large                                                 24.49/iter    +/- 0.06
    f32::div_euclid_medium                                                 8.44/iter    +/- 0.19
    f32::div_euclid_small                                                  4.02/iter    +/- 0.01
    f64::div_euclid_large                                                151.79/iter    +/- 0.24
    f64::div_euclid_medium                                                10.56/iter    +/- 0.02
    f64::div_euclid_small                                                  4.16/iter    +/- 0.02
```

After the change (correctly rounded):
```
    f128::div_euclid_large                                               565.23/iter     +/- 1.73
    f128::div_euclid_medium                                              414.26/iter     +/- 1.92
    f128::div_euclid_small                                                11.54/iter     +/- 0.01
    f16::div_euclid_large                                                 11.90/iter     +/- 0.45
    f16::div_euclid_small                                                 11.87/iter     +/- 0.42
    f32::div_euclid_large                                                  5.14/iter     +/- 0.36
    f32::div_euclid_medium                                                 4.97/iter     +/- 0.18
    f32::div_euclid_small                                                  4.26/iter     +/- 0.23
    f64::div_euclid_large                                                  8.04/iter     +/- 0.42
    f64::div_euclid_medium                                                 7.19/iter     +/- 0.39
    f64::div_euclid_small                                                  4.75/iter     +/- 0.30
```

So the corrected implementation actually improves performance significantly in many cases, and there is no significant performance regression.

tczajka committed:
div_euclid: add a fast path for overflow

tczajka commented:
> This PR from @tczajka takes the softfloat implementation approach.

The current (old) implementation is also really soft-float. The operation is not built into hardware. Additionally, it uses `%` (`fmod`) internally which is not only not implemented in hardware, but it is also very slow for large numbers.

My implementation is **one** integer division with some quick adjustments for correct rounding. This is why you see large performance gains.

One thing that has room for optimization is `f128`. I need 256-bit operations for this, so I use a custom implementation of `U256` with a naive division algorithm. Its performance can easily be improved, either by a change in the algorithm, or by using LLVM's built-in `i256` type.

workingjubilee commented:
> **+2,219** −65

oh dear

tczajka commented:
The code for `f32` and `f64` is identical, the code for `f16` and `f128` is a bit different. So I'm open to suggestions for how to best write this with minimal duplication. Macros for `f32` and `f64`?

tgross35 commented:
I still think that a better place for more complex soft float routines is probably in `libm` to be distributed as part of `compiler-builtins` - we already have the [float traits](https://github.com/rust-lang/libm/blob/aa3d242bde84f94858387e888436fc536d9246fb/src/math/support/float_traits.rs), [`u256`](https://github.com/rust-lang/compiler-builtins/blob/520e2a09ecf871d10193b91e27a547bbc7924c93/src/int/big.rs) (currently in compiler-builtins and doesn't have division, but that can be changed), tests against MPFR, and (soon) [exhaustive tests](https://github.com/rust-lang/libm/pull/388). This isn't technically a math.h routine but we don't need to make it public in `libm` itself, just for the reexport in builtins.

We don't currently have anything like this though, @Amanieu would there be any concerns here?

bors commented:
:umbrella: The latest upstream changes (presumably #135947) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

workingjubilee commented on library/std/src/f128.rs:
```diff
@@ -235,10 +237,14 @@ impl f128 {
     /// Calculates Euclidean division, the matching method for `rem_euclid`.
     ///
-    /// This computes the integer `n` such that
+    /// In infinite precision this computes the integer `n` such that
```
Usually I see this arranged as "This computes the operation on the integer `n` with infinite precision"

workingjubilee replied:
Those are fine too. It's honestly at least partly that "In infinite precision" feels like an odd way to start the sentence.

workingjubilee commented on library/std/src/f128.rs:
```diff
@@ -235,10 +237,14 @@ impl f128 {
+    /// In other words, the result is `self / rhs` rounded to the
+    /// integer `n` such that `self >= n * rhs`.
+    ///
+    /// However, due to a floating point round-off error the result can be rounded if
+    /// `self` is much larger in magnitude than `rhs`, such that `n` can't be represented
+    /// exactly.
```
This makes the "with infinite precision" sound incorrect. It is not a random "round-off error", the rounding is a consequence of fitting the result into the floating-point value's representation. It is questionable to call it an "error" at all, since it is a deliberate choice in how floats work. Please be clear when identifying such. We can afford a few words.

workingjubilee commented on library/std/src/f128.rs:
```diff
@@ -264,29 +270,23 @@ impl f128 {
-    /// closest floating point number in the real numbers and thus fulfills the
-    /// property `self == self.div_euclid(rhs) * rhs + self.rem_euclid(rhs)`
-    /// approximately.
+    /// In infinite precision the return value `r` satisfies `0 <= r < rhs.abs()`.
+    ///
+    /// However, due to a floating point round-off error the result can round to
```
same comments

workingjubilee commented on library/std/src/f128.rs:
```diff
@@ -264,29 +270,23 @@ impl f128 {
     /// The result of this operation is guaranteed to be the rounded
     /// infinite-precision result.
     ///
+    /// The result is precise when `self >= 0.0`.
```
Floating-point operations are always precise (in the sense that they are always precisely conforming to floating-point semantics), they are not always _accurate_, see https://en.wikipedia.org/wiki/Accuracy_and_precision for more on the distinction I am drawing.
```suggestion
    /// The result is exactly accurate when `self >= 0.0`.
```

We could also say within 0 ULP, I suppose?

workingjubilee replied:
The "exactly representable" form sounds good.

workingjubilee commented on library/std/src/f16.rs:
The documentation-related comments I have on f128.rs apply here.

workingjubilee commented on library/std/src/f32.rs:
The documentation-related comments I have on f128.rs apply here.

workingjubilee commented on library/std/src/f64.rs:
The documentation-related comments I have on f128.rs apply here.

workingjubilee reviewed (changes requested):
Finally getting around to this. I have some documentation-related comments for a first pass since with that concern out of the way the rest of this is much less, relatively speaking, to review.

workingjubilee commented on library/std/src/f128/soft/tests.rs:
```diff
@@ -0,0 +1,49 @@
+
+    assert_eq!(Representation::new(f128::MIN_POSITIVE), Representation {
+        sign: Sign::Positive,
+        abs: Positive::Finite(PositiveFinite { exp: -16494, mantissa: 1 << 112 }),
+    });
+
+    assert_eq!(Representation::new(f128::MIN_POSITIVE / 2.0), Representation {
+        sign: Sign::Positive,
+        abs: Positive::Finite(PositiveFinite { exp: -16495, mantissa: 1 << 112 }),
```
This `1 << 112` is a meaningful magic number, and should be derivable from the existing constants (`SIGNIFICAND_BITS` and so on).

Can we lift this into a `const MANTISSA_TOP_BIT` (or whatever) that computes it? If you want to also `assert_eq!` that const against `1 << 112` still, that's fine too.

workingjubilee commented on library/std/src/f128/soft/tests.rs:
```diff
@@ -0,0 +1,49 @@
+
+    assert_eq!(Representation::new(f128::MAX), Representation {
+        sign: Sign::Positive,
+        abs: Positive::Finite(PositiveFinite { exp: 16271, mantissa: (1 << 113) - 1 }),
```
likewise `const MANTISSA_MAX`

workingjubilee commented on library/std/src/f32/soft/tests.rs:
```diff
@@ -0,0 +1,49 @@
+
+    assert_eq!(Representation::new(f32::MIN_POSITIVE), Representation {
+        sign: Sign::Positive,
+        abs: Positive::Finite(PositiveFinite { exp: -149, mantissa: 0x800000 }),
+    });
+
+    assert_eq!(Representation::new(f32::MIN_POSITIVE / 2.0), Representation {
+        sign: Sign::Positive,
+        abs: Positive::Finite(PositiveFinite { exp: -150, mantissa: 0x800000 }),
+    });
+
+    assert_eq!(Representation::new(f32::MAX), Representation {
+        sign: Sign::Positive,
+        abs: Positive::Finite(PositiveFinite { exp: 104, mantissa: 0xffffff }),
```
This is a different way to write these mantissa values than is done elsewhere...?

workingjubilee commented on library/std/src/f128/soft.rs:
```diff
@@ -0,0 +1,358 @@
+}
+
+/// Euclidean division.
+#[allow(dead_code)]
```
Why this `allow`...?

workingjubilee commented on library/std/src/f64/soft/tests.rs:
```diff
@@ -0,0 +1,49 @@
+
+    assert_eq!(Representation::new(f64::MIN_POSITIVE), Representation {
+        sign: Sign::Positive,
+        abs: Positive::Finite(PositiveFinite { exp: -1074, mantissa: 1 << 52 }),
```
...this exponent is a fair bit away from `-1021`, so I am going to have to go over  a number of things very carefully, and I am not immediately seeing an explanation of the encoding choice in this module. I'd like to see a comment somewhere (perhaps on the struct, or on the associated constants, or this test?), that explains why what I am seeing here is Actually Correct.

workingjubilee commented on library/std/src/f64/soft.rs:
```diff
@@ -0,0 +1,352 @@
+/// Encoded exponent = EXPONENT_BIAS + actual exponent.
+const EXPONENT_BIAS: Exponent = 2 - FP::MIN_EXP;
+
+/// Represents an `FP` number.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct Representation {
+    /// Sign.
+    sign: Sign,
+    /// Absolute value.
+    abs: Positive,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Sign {
+    Positive,
+    Negative,
+}
+
+/// Represents a positive number.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Positive {
+    Zero,
+    Finite(PositiveFinite),
+    Infinity,
+    NaN,
+}
+
+/// Represents a non-zero, positive finite number.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct PositiveFinite {
+    /// `number = 2^exp * mantissa`.
+    exp: Exponent,
+    /// `2^SIGNIFICAND_BITS <= mantissa < 2^(SIGNIFICAND_BITS + 1)`
+    mantissa: Bits,
+}
```
This PR winds up repeating these types a number of times... can it just be made generic? If we need to have separate `impl` blocks for each concrete combination of params, so be it, but we could at least have a unified declaration in an adjacent `pub(crate) mod float;`

workingjubilee reviewed (changes requested):
Another partial review, this time of the soft-float tests. Their current state points to a need for slightly more documentation and perhaps a slightly different organization of the code proper. The way things currently are, I do not quite agree with any assessment that sounds like "well, you only need to review one type's implementation", since I need to make sure that there isn't some important deviation.

workingjubilee commented:
@tczajka If you can make the soft-float impl as generic as possible given the current state of things then I think we can very quickly actually finish this out.

tgross35 commented:
> @tczajka If you can make the soft-float impl as generic as possible given the current state of things then I think we can very quickly actually finish this out.

Pls can we consider `libm` rather than adding float traits in yet another place

workingjubilee commented:
"generics" is not necessarily "trait".

workingjubilee commented:
and libm will not be able to have const traits until those are completely finished.

workingjubilee commented:
and as already discussed, this PR would probably benefit if it had access to the powers of the compiler (and the compiler was augmented).

workingjubilee commented:
Just because libs-api constantly bats their eyes and asks "what's the use-case?" over "why make floats generic?" doesn't mean there is not one, and it is frankly absurd that it occurs outside the standard library.

workingjubilee commented:
...apologies, just...

@tgross35 I am curious to know why, to you, "adding float traits in yet another place" is a sign that this PR should be against another repo, instead of fixing the fact that we have "float traits" everywhere except in the core libraries?

workingjubilee commented:
It cannot be "because it would be a miserable pile of bikeshedding". We're just inflicting a miserable pile of bikeshedding on everyone every time they have to reimpl their own pile of float traits. Skipping numeric traits in std made sense in 2015 but it doesn't really make the same sense in 2025.

workingjubilee commented:
It also does not actually seem to help us to have three different repositories one must consult to verify that a floating point implementation that we use is correct, either.

tgross35 commented:
> @tgross35 I am curious to know why, to you, "adding float traits in yet another place" is a sign that this PR should be against another repo, instead of fixing the fact that we have "float traits" everywhere except in the core libraries?

I don't disagree that it would be very nice to have the traits in `core`. Keeping traits in one place was just one motivation in response to generics. The reasoning is deeper though: `libm` does have the traits (for floats, integers, doubling width, etc) but it also has a reasonably thorough test suite (tests against MPFR with edge cases, fuzzing, extensive/exhaustive tests), icount benchmarks, `u256`, and a handful of other code reuse things. Not necessarily that this routine would add _that_ much more, imo we just have a better home for softfloat :)

> as already discussed, this PR would probably benefit if it had access to the powers of the compiler (and the compiler was augmented).

I'm not seeing this, got a link?

> and libm will not be able to have const traits until those are completely finished.

We do have a mechanism for using nightly features with a fallback. But we may not need to make these const for them to be const if they can become intrinsics (from some discussion with const-eval I'm planning to use the `libm` generic methods to add an implementation over `rustc_apfloat` types, which would allow us to just use `libm` in CTFE).

> It also does not actually seem to help us to have three different repositories one must consult to verify that a floating point implementation that we use is correct, either.

Yeah, it's not ideal. I would like to merge `libm` into `compiler-builtins` but that has unfortunately been blocked for months on https://github.com/rust-lang/rust/pull/135501 (well, just making it a subtree isn't, but some of the restructuring I would like to do afterward is). After that we could probably bump to a subtree here, TBD.

workingjubilee commented:
> > as already discussed, this PR would probably benefit if it had access to the powers of the compiler (and the compiler was augmented).
>
> I'm not seeing this, got a link?

I mean the mention of using a `u256` provided directly by the compiler: there's nothing technically infeasible about us doing so, given that LLVM supports it.

workingjubilee commented:
Hm. I am not sure some of those tests should not be in libcore, but I do not expect us to add a probably-randomization-driven test suite to our already flakey CI, for reasons previously discussed, so if that is what is ideal then moving this over there is fine.

Just as long as it's not about our 99 float traits lying around.

tczajka commented:
Sorry I've been busy with other things, I'm planning to get back to the review within ~2 weeks.

alex-semenyuk commented:
@tczajka
Thanks for your contribution.
Did you have a chance to take a look
